Batteries/Data/Fin/Basic.lean

```diff
@@ -17,3 +17,69 @@ alias enum := Array.finRange
 
 @[deprecated (since := "2024-11-15")]
 alias list := List.finRange
+
+variable (n : Nat) (α : Fin (n + 1) → Sort _)
+
+/-- Dependent version of `Fin.foldl`. -/
+@[inline] def fold (f : ∀ (i : Fin n), α i.castSucc → α i.succ) (init : α 0) : α (last n) :=
+  loop 0 (Nat.zero_lt_succ n) init where
+  /-- Inner loop for `Fin.fold`. `Fin.fold.loop n α f i h x = f n (f (n-1) (... (f i x)))`  -/
+  loop (i : Nat) (h : i < n + 1) (x : α ⟨i, h⟩) : α (last n) :=
+    if h' : i < n then
+      loop (i + 1) (Nat.succ_lt_succ h') (f ⟨i, h'⟩ x)
+    else
+      haveI : ⟨i, h⟩ = last n := by ext; simp; omega
+      _root_.cast (congrArg α this) x
+
+/-- Dependent version of `Fin.foldr`. -/
+@[inline] def foldRev (f : ∀ (i : Fin n), α i.succ → α i.castSucc) (init : α (last n)) : α 0 :=
+  loop n (Nat.lt_succ_self n) init where
+  /-- Inner loop for `Fin.foldRev`.
+    `Fin.foldRev.loop n α f i h x = f 0 (f 1 (... (f i x)))`  -/
+  loop (i : Nat) (h : i < n + 1) (x : α ⟨i, h⟩) : α 0 :=
+    if h' : i > 0 then
+      loop (i - 1) (by omega) (f ⟨i - 1, by omega⟩
+        (by simpa [Nat.sub_one_add_one_eq_of_pos h'] using x))
+    else
+      haveI : ⟨i, h⟩ = 0 := by ext; simp; omega
+      _root_.cast (congrArg α this) x
+
+/-- Dependent version of `Fin.foldlM`. -/
+@[inline] def foldM [Monad m] (f : ∀ (i : Fin n), α i.castSucc → m (α i.succ)) (init : α 0)
```

Comment:
```suggestion
@[specialize] def foldM [Monad m] (f : ∀ (i : Fin n), α i.castSucc → m (α i.succ)) (init : α 0)
```

Reply:
What is the difference between `inline` and `specialize`?

Reply:
Answered in another comment.